### ISSUE TITLE
Feature proposal - Affinities

### ISSUE THREAD

**Issue by yellowmegaman** [hashicorp/nomad#2509]
I'd like to understand how to achieve wider interpretation in constraints.

Use case:
```
constraint {
  attribute = "${attr.unique.hostname}"
  regexp = "media"
}
```
But what happens when host media is down or drained? What i'd want - to have "preferred hosts" and ability to launch the app/container elsewhere if host is suddenly down down.

like:

```
regexp = value
regexp_preferred = value
```

Thanks a lot in advance!

**Comment by jippi** [comment]
I guess a feature like this would make sense once allocation rebalance / force allocation movement is in place :)

**Comment by dadgar** [comment]
Hey this would be solved with affinities and not constraints. You would have two affinities, one for the preferred with a high weight and and then the secondary value with a lower weight. I am going to rename the issue.

**Comment by yellowmegaman** [comment]
Awesome! Thanks for clarification. Any milestone it can be attached to?

**Comment by dadgar** [comment]
@yellowmegaman Not at this time. It will not be 0.6.0 and likely not 0.7 so some time after that.

**Comment by yellowmegaman** [comment]
Well, patience is golden, it's in plans and that's great.

**Comment by yellowmegaman** [comment]
Since creating this issue, I've reworked all containers to be host-independent, even databases. But not load balancers. 

But still, sometimes i wish i could place priority for databases to be on higher IOPS instance/geo stuff without making strict constraint.

**Comment by jippi** [comment]
See #4513 :)

**Comment by yellowmegaman** [comment]
@jippi thanks!)

**Comment by dantoml** [comment]
Hey folks, Affinities have landed as part of the 0.9 betas - You can find documentation [here](https://www.nomadproject.io/docs/job-specification/affinity.html). 😄
